File: src/vault/store.ts
import { randomBytes, randomUUID } from "node:crypto";
import { chmod, lstat, mkdir, readFile, readdir, rename, rm, writeFile } from "node:fs/promises";
import path from "node:path";
import { DEFAULT_KDF_PARAMETERS, deriveKey } from "../crypto/kdf.js";
import { unwrapDataKey, wrapDataKey } from "../crypto/key-wrap.js";
import { ensurePrivateDirectory, MANIFEST_VERSION, runtimeRoot, vaultRoot } from "../config.js";
import { AuthenticationError, VaultError } from "../errors.js";
import type { KdfParameters, UserManifest } from "../types.js";
import { acquireUserLock } from "../runtime/lock.js";
import { sealDirectory, unsealDirectory } from "./archive.js";

const USERNAME_PATTERN = /^[a-zA-Z0-9][a-zA-Z0-9._-]{0,63}$/;
const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

function userDirectoryName(username: string): string {
  return username.toLowerCase();
}

function manifestAssociatedData(manifest: Omit<UserManifest, "wrappedDataKey">): Buffer {
  return Buffer.from(
    JSON.stringify({
      context: "codex-vault-key-wrap-v1",
      version: manifest.version,
      id: manifest.id,
      username: manifest.username,
      createdAt: manifest.createdAt,
      kdf: manifest.kdf,
      vaultFormat: manifest.vaultFormat,
    }),
    "utf8",
  );
}

async function writePrivateJson(filePath: string, value: unknown): Promise<void> {
  await writeFile(filePath, `${JSON.stringify(value, null, 2)}\n`, {
    encoding: "utf8",
    flag: "wx",
    mode: 0o600,
  });
  await chmod(filePath, 0o600);
}

export interface UnlockedVault {
  username: string;
  codexHome: string;
  close(): Promise<void>;
}

export class VaultStore {
  constructor(private readonly root = vaultRoot()) {}

  async initialize(): Promise<void> {
    await ensurePrivateDirectory(this.root);
    await ensurePrivateDirectory(this.usersRoot());
  }

  async addUser(username: string, password: string): Promise<void> {
    this.validateUsername(username);
    this.validatePassword(password);
    await this.initialize();

    const userRoot = this.userRoot(username);
    try {
      await lstat(userRoot);
      throw new VaultError(`User already exists: ${username}`);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== "ENOENT") {
        throw error;
      }
    }
    const staging = `${userRoot}.creating-${randomUUID()}`;
    await mkdir(staging, { recursive: false, mode: 0o700 });

    const dataKey = randomBytes(32);
    const salt = randomBytes(16);
    const kdf: KdfParameters = {
      ...DEFAULT_KDF_PARAMETERS,
      salt: salt.toString("base64"),
    };
    const wrappingKey = await deriveKey(password, kdf);

    try {
      const manifestHeader: Omit<UserManifest, "wrappedDataKey"> = {
        version: MANIFEST_VERSION,
        id: randomUUID(),
        username,
        createdAt: new Date().toISOString(),
        kdf,
        vaultFormat: "cdxvlt01",
      };
      const manifest: UserManifest = {
        ...manifestHeader,
        wrappedDataKey: wrapDataKey(
          dataKey,
          wrappingKey,
          manifestAssociatedData(manifestHeader),
        ),
      };

      await writePrivateJson(path.join(staging, "manifest.json"), manifest);
      const emptyState = path.join(staging, "empty-state");
      await mkdir(emptyState, { mode: 0o700 });
      await sealDirectory(emptyState, path.join(staging, "state.cvlt"), dataKey);
      await rm(emptyState, { recursive: true, force: true });
      await mkdir(path.dirname(userRoot), { recursive: true, mode: 0o700 });
      await rename(staging, userRoot);
    } catch (error) {
      await rm(staging, { recursive: true, force: true });
      throw error;
    } finally {
      dataKey.fill(0);
      wrappingKey.fill(0);
    }
  }

  async listUsers(): Promise<string[]> {
    await this.initialize();
    const entries = await readdir(this.usersRoot(), { withFileTypes: true });
    const users: string[] = [];

    for (const entry of entries) {
      if (!entry.isDirectory() || entry.name.includes(".creating-")) {
        continue;
      }
      try {
        const manifest = await this.readManifest(path.join(this.usersRoot(), entry.name));
        users.push(manifest.username);
      } catch {
        // Ignore incomplete or invalid directories when listing accounts.
      }
    }

    return users.sort((a, b) => a.localeCompare(b));
  }

  async unlock(username: string, password: string): Promise<UnlockedVault> {
    this.validateUsername(username);
    const userRoot = this.userRoot(username);
    let releaseLock: () => Promise<void> = async () => {};
    let runtimeDirectory: string | undefined;
    let dataKey: Buffer | undefined;
    let wrappingKey: Buffer | undefined;

    try {
      const userMetadata = await lstat(userRoot);
      if (!userMetadata.isDirectory() || userMetadata.isSymbolicLink()) {
        throw new AuthenticationError("Invalid username or password");
      }
      releaseLock = await acquireUserLock(path.join(userRoot, "active.lock"));
      const manifest = await this.readManifest(userRoot);
      if (manifest.username.toLowerCase() !== username.toLowerCase()) {
        throw new AuthenticationError("Invalid username or password");
      }

      wrappingKey = await deriveKey(password, manifest.kdf);
      const { wrappedDataKey: _wrappedDataKey, ...manifestHeader } = manifest;
      dataKey = unwrapDataKey(
        manifest.wrappedDataKey,
        wrappingKey,
        manifestAssociatedData(manifestHeader),
      );

      const baseRuntime = await runtimeRoot();
      runtimeDirectory = path.join(baseRuntime, `${manifest.id}-${randomUUID()}`);
      const codexHome = path.join(runtimeDirectory, "codex-home");
      await mkdir(codexHome, { recursive: true, mode: 0o700 });
      await unsealDirectory(path.join(userRoot, "state.cvlt"), codexHome, dataKey);

      let closed = false;
      return {
        username: manifest.username,
        codexHome,
        close: async () => {
          if (closed) {
            return;
          }
          closed = true;
          try {
            await sealDirectory(codexHome, path.join(userRoot, "state.cvlt"), dataKey!);
          } finally {
            await rm(runtimeDirectory!, { recursive: true, force: true });
            dataKey!.fill(0);
            wrappingKey!.fill(0);
            await releaseLock();
          }
        },
      };
    } catch (error) {
      if (runtimeDirectory) {
        await rm(runtimeDirectory, { recursive: true, force: true });
      }
      dataKey?.fill(0);
      wrappingKey?.fill(0);
      await releaseLock();
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        throw new AuthenticationError("Invalid username or password");
      }
      throw error;
    }
  }

  private usersRoot(): string {
    return path.join(this.root, "users");
  }

  private userRoot(username: string): string {
    return path.join(this.usersRoot(), userDirectoryName(username));
  }

  private validateUsername(username: string): void {
    if (!USERNAME_PATTERN.test(username)) {
      throw new VaultError(
        "Username must be 1-64 characters and contain only letters, numbers, dot, underscore, or hyphen",
      );
    }
  }

  private validatePassword(password: string): void {
    if (password.length < 12) {
      throw new VaultError("Password must contain at least 12 characters");
    }
  }

  private async readManifest(userRoot: string): Promise<UserManifest> {
    const manifestPath = path.join(userRoot, "manifest.json");
    const metadata = await lstat(manifestPath);
    if (!metadata.isFile() || metadata.isSymbolicLink() || metadata.size > 64 * 1024) {
      throw new VaultError("Invalid user manifest file");
    }
    const parsed = JSON.parse(await readFile(manifestPath, "utf8")) as Partial<UserManifest> | null;
    if (
      !parsed ||
      parsed.version !== MANIFEST_VERSION ||
      typeof parsed.id !== "string" ||
      !UUID_PATTERN.test(parsed.id) ||
      typeof parsed.username !== "string" ||
      !USERNAME_PATTERN.test(parsed.username) ||
      typeof parsed.createdAt !== "string" ||
      !Number.isFinite(Date.parse(parsed.createdAt)) ||
      parsed.kdf?.algorithm !== "argon2id" ||
      typeof parsed.kdf.salt !== "string" ||
      typeof parsed.kdf.memoryCostKiB !== "number" ||
      typeof parsed.kdf.timeCost !== "number" ||
      typeof parsed.kdf.parallelism !== "number" ||
      typeof parsed.kdf.outputLength !== "number" ||
      parsed.wrappedDataKey?.algorithm !== "aes-256-gcm" ||
      typeof parsed.wrappedDataKey.iv !== "string" ||
      typeof parsed.wrappedDataKey.ciphertext !== "string" ||
      typeof parsed.wrappedDataKey.authenticationTag !== "string" ||
      parsed.vaultFormat !== "cdxvlt01"
    ) {
      throw new VaultError("Unsupported or invalid user manifest");
    }
    return parsed as UserManifest;
  }
}

File: src/store.rs
use std::{
    fs::{self, File, OpenOptions},
    io::{Read, Seek, SeekFrom, Write},
    path::{Path, PathBuf},
};

#[cfg(unix)]
use std::os::unix::fs::OpenOptionsExt;

use fs2::FileExt;
use tempfile::TempDir;
use uuid::Uuid;
use zeroize::Zeroizing;

use crate::{
    archive::{seal_directory, unseal_directory},
    config::{
        default_vault_root, ensure_private_directory, select_runtime_root,
        validate_private_directory, validate_private_file,
    },
    crypto::{default_kdf_parameters, derive_key, unwrap_data_key, wrap_data_key},
    error::{Result, VaultError},
    manifest::{UserManifest, is_valid_username},
};

const MANIFEST_SIZE_LIMIT: u64 = 64 * 1024;

/// Persistent encrypted Codex profile storage.
pub struct VaultStore {
    root: PathBuf,
    runtime_root: Option<PathBuf>,
}

impl VaultStore {
    /// Creates a store from `CODEX_VAULT_HOME` or the current user's home directory.
    ///
    /// # Errors
    ///
    /// Returns an error when neither storage location is available.
    pub fn from_environment() -> Result<Self> {
        Ok(Self::new(default_vault_root()?))
    }

    #[must_use]
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self {
            root: root.into(),
            runtime_root: None,
        }
    }

    #[must_use]
    pub fn with_runtime_root(mut self, runtime_root: impl Into<PathBuf>) -> Self {
        self.runtime_root = Some(runtime_root.into());
        self
    }

    /// Creates the private storage directories and enforces restrictive permissions.
    ///
    /// # Errors
    ///
    /// Returns an error when the directory is unsafe, owned by another UID, or inaccessible.
    pub fn initialize(&self) -> Result<()> {
        ensure_private_directory(&self.root)?;
        ensure_private_directory(&self.users_root())
    }

    /// Creates an encrypted application-level user profile.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid credentials, duplicate users, random-source failures, or I/O
    /// failures while creating the initial vault.
    pub fn add_user(&self, username: &str, password: &[u8]) -> Result<()> {
        validate_username(username)?;
        validate_password(password)?;
        self.initialize()?;

        let user_root = self.user_root(username);
        if fs::symlink_metadata(&user_root).is_ok() {
            return Err(VaultError::UserExists(username.into()));
        }

        let staging = self
            .users_root()
            .join(format!(".creating-{}", Uuid::new_v4()));
        ensure_private_directory(&staging)?;

        let result = (|| {
            let kdf = default_kdf_parameters()?;
            let wrapping_key = derive_key(password, &kdf)?;
            let mut data_key = Zeroizing::new([0_u8; 32]);
            getrandom::fill(data_key.as_mut())
                .map_err(|_| VaultError::Cryptography("operating-system random source failed"))?;
            let mut manifest = UserManifest::new_header(username.into(), kdf)?;
            manifest.wrapped_data_key =
                wrap_data_key(&data_key, &wrapping_key, &manifest.associated_data()?)?;

            write_private_json(&staging.join("manifest.json"), &manifest)?;
            let empty = tempfile::tempdir()?;
            seal_directory(empty.path(), &staging.join("state.cvlt"), &data_key)?;
            fs::rename(&staging, &user_root)?;
            Ok(())
        })();

        if result.is_err() {
            let _ = fs::remove_dir_all(&staging);
        }
        result
    }

    /// Lists valid user profiles without unlocking their encrypted state.
    ///
    /// # Errors
    ///
    /// Returns an error when the store cannot be initialized or enumerated.
    pub fn list_users(&self) -> Result<Vec<String>> {
        self.initialize()?;
        let mut users = Vec::new();
        for entry in fs::read_dir(self.users_root())? {
            let entry = entry?;
            if !entry.file_type()?.is_dir() || entry.file_name().to_string_lossy().starts_with('.')
            {
                continue;
            }
            if let Ok(manifest) = read_manifest(&entry.path()) {
                users.push(manifest.username);
            }
        }
        users.sort_unstable();
        Ok(users)
    }

    /// Authenticates a user and extracts their encrypted Codex state into a private runtime.
    ///
    /// # Errors
    ///
    /// Returns an error for failed authentication, concurrent access, invalid encrypted state, or
    /// runtime-directory failures.
    pub fn unlock(&self, username: &str, password: &[u8]) -> Result<UnlockedVault> {
        validate_username(username)?;
        let user_root = self.user_root(username);
        validate_private_directory(&user_root).map_err(|_| VaultError::Authentication)?;
        let mut lock = Some(acquire_lock(&user_root.join("active.lock"))?);

        let result = (|| {
            let manifest = read_manifest(&user_root)?;
            manifest.validate()?;
            if !manifest.username.eq_ignore_ascii_case(username) {
                return Err(VaultError::Authentication);
            }
            let wrapping_key =
                derive_key(password, &manifest.kdf).map_err(|_| VaultError::Authentication)?;
            let data_key = unwrap_data_key(
                &manifest.wrapped_data_key,
                &wrapping_key,
                &manifest.associated_data()?,
            )?;

            let runtime_root = match &self.runtime_root {
                Some(path) => {
                    ensure_private_directory(path)?;
                    path.clone()
                }
                None => select_runtime_root()?,
            };
            let runtime = tempfile::Builder::new()
                .prefix(&format!("{}-", manifest.id))
                .tempdir_in(runtime_root)?;
            let codex_home = runtime.path().join("codex-home");
            ensure_private_directory(&codex_home)?;
            unseal_directory(&user_root.join("state.cvlt"), &codex_home, &data_key)?;

            Ok(UnlockedVault {
                username: manifest.username,
                codex_home,
                runtime,
                state_path: user_root.join("state.cvlt"),
                data_key,
                lock: lock
                    .take()
                    .ok_or(VaultError::Cryptography("vault lock was not available"))?,
            })
        })();

        if let (Err(_), Some(lock)) = (&result, lock.as_ref()) {
            let _ = FileExt::unlock(lock);
        }
        result
    }

    fn users_root(&self) -> PathBuf {
        self.root.join("users")
    }

    fn user_root(&self, username: &str) -> PathBuf {
        self.users_root().join(username.to_ascii_lowercase())
    }
}

/// An authenticated, exclusively locked Codex profile with a temporary plaintext runtime.
pub struct UnlockedVault {
    username: String,
    codex_home: PathBuf,
    runtime: TempDir,
    state_path: PathBuf,
    data_key: Zeroizing<[u8; 32]>,
    lock: File,
}

impl UnlockedVault {
    #[must_use]
    pub fn username(&self) -> &str {
        &self.username
    }

    #[must_use]
    pub fn codex_home(&self) -> &Path {
        &self.codex_home
    }

    /// Reseals the runtime state, releases the lock, and removes the temporary directory.
    ///
    /// # Errors
    ///
    /// Returns an error when resealing or releasing the file lock fails. The temporary directory is
    /// still removed to avoid leaving plaintext behind.
    pub fn close(self) -> Result<()> {
        let seal_result = seal_directory(&self.codex_home, &self.state_path, &self.data_key);
        let unlock_result = FileExt::unlock(&self.lock).map_err(VaultError::Io);
        drop(self.runtime);
        seal_result?;
        unlock_result
    }
}

fn validate_username(username: &str) -> Result<()> {
    if !is_valid_username(username) {
        return Err(VaultError::InvalidConfiguration(
            "username must be 1-64 ASCII letters, numbers, dot, underscore, or hyphen".into(),
        ));
    }
    Ok(())
}

fn validate_password(password: &[u8]) -> Result<()> {
    if password.len() < 12 {
        return Err(VaultError::InvalidConfiguration(
            "password must contain at least 12 bytes".into(),
        ));
    }
    Ok(())
}

fn write_private_json(path: &Path, value: &UserManifest) -> Result<()> {
    let mut options = OpenOptions::new();
    options.write(true).create_new(true);
    #[cfg(unix)]
    options.mode(0o600);
    let mut file = options.open(path)?;
    serde_json::to_writer_pretty(&mut file, value)?;
    file.write_all(b"\n")?;
    file.sync_all()?;
    Ok(())
}

fn read_manifest(user_root: &Path) -> Result<UserManifest> {
    let path = user_root.join("manifest.json");
    let metadata = validate_private_file(&path)?;
    if metadata.len() > MANIFEST_SIZE_LIMIT {
        return Err(VaultError::InvalidConfiguration(
            "invalid user manifest file".into(),
        ));
    }
    let mut options = OpenOptions::new();
    options.read(true);
    #[cfg(unix)]
    options.custom_flags(nix::libc::O_NOFOLLOW);
    let file = options.open(path)?;
    let capacity = usize::try_from(metadata.len()).map_err(|_| {
        VaultError::InvalidConfiguration("user manifest size does not fit this platform".into())
    })?;
    let mut contents = Vec::with_capacity(capacity);
    file.take(MANIFEST_SIZE_LIMIT + 1)
        .read_to_end(&mut contents)?;
    if contents.len() as u64 > MANIFEST_SIZE_LIMIT {
        return Err(VaultError::InvalidConfiguration(
            "user manifest exceeds size limit".into(),
        ));
    }
    let manifest: UserManifest = serde_json::from_slice(&contents)?;
    manifest.validate()?;
    Ok(manifest)
}

fn acquire_lock(path: &Path) -> Result<File> {
    let mut options = OpenOptions::new();
    options.read(true).write(true).create(true);
    #[cfg(unix)]
    {
        options.mode(0o600);
        options.custom_flags(nix::libc::O_NOFOLLOW);
    }
    let mut file = options.open(path)?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        file.set_permissions(fs::Permissions::from_mode(0o600))?;
    }
    match FileExt::try_lock_exclusive(&file) {
        Ok(()) => {}
        Err(error) if error.kind() == std::io::ErrorKind::WouldBlock => {
            return Err(VaultError::Locked);
        }
        Err(error) => return Err(VaultError::Io(error)),
    }
    file.set_len(0)?;
    file.seek(SeekFrom::Start(0))?;
    writeln!(file, "{{\"pid\":{}}}", std::process::id())?;
    file.flush()?;
    Ok(file)
}
